heft-plugins/heft-webpack4-plugin/src/Webpack4Plugin.ts

```diff
@@ -0,0 +1,283 @@
+    taskSession: IHeftTaskSession,
+    heftConfiguration: HeftConfiguration,
+    options: IWebpack4PluginOptions
+  ): void {
+    // These get set in the run hook and used in the onConfigureWebpackHook
+    let production: boolean;
+    let serveMode: boolean;
+    let watchMode: boolean;
+
+    // Provide the initial webpack configuration by tapping the hook
+    this.accessor.onConfigureWebpackHook.tapPromise(
+      PLUGIN_NAME,
+      async (existingConfiguration: IWebpackConfiguration | false) => {
+        if (existingConfiguration) {
+          taskSession.logger.terminal.writeVerboseLine(
+            'Skipping loading webpack config file because the webpack config has already been set.'
+          );
+          return existingConfiguration;
+        } else {
+          const configurationLoader: WebpackConfigurationLoader = new WebpackConfigurationLoader(
+            taskSession.logger,
+            production,
+            serveMode
+          );
+          const webpack: typeof TWebpack = await this._getWebpackAsync();
+          return await configurationLoader.tryLoadWebpackConfigurationAsync({
+            ...options,
+            taskSession,
+            heftConfiguration,
+            webpack
+          });
+        }
+      }
+    );
+
+    taskSession.hooks.clean.tapPromise(PLUGIN_NAME, async (cleanOptions: IHeftTaskCleanHookOptions) => {
+      // Obtain the finalized webpack configuration
+      const webpackConfiguration: IWebpackConfiguration | null = await this._getWebpackConfigurationAsync();
+      if (webpackConfiguration) {
+        const webpackConfigurationArray: IWebpackConfigurationWithDevServer[] =
+          Array.isArray(webpackConfiguration) ? webpackConfiguration : [webpackConfiguration];
+
+        // Add each output path to the clean list
+        for (const config of webpackConfigurationArray) {
+          if (config.output?.path) {
+            cleanOptions.addDeleteOperations({ sourcePath: config.output.path });
+          }
+        }
+      }
+    });
+
+    taskSession.hooks.run.tapPromise(PLUGIN_NAME, async (runOptions: IHeftTaskRunHookOptions) => {
+      production = runOptions.production;
+      // TODO: Support watch mode
+      watchMode = false;
+      // TODO: Support serve mode
+      serveMode = false;
+
+      // Run webpack with the finalized webpack configuration
+      const webpackConfiguration: IWebpackConfiguration | null = await this._getWebpackConfigurationAsync();
+      await this._runWebpackAsync(
+        taskSession,
+        heftConfiguration,
+        webpackConfiguration,
+        serveMode,
+        watchMode
+      );
+    });
+  }
+
+  private async _getWebpackAsync(): Promise<typeof TWebpack> {
+    if (!this._webpack) {
+      this._webpack = await import('webpack');
+    }
+    return this._webpack;
+  }
+
+  private async _getWebpackConfigurationAsync(): Promise<IWebpackConfiguration | undefined> {
+    if (this._webpackConfiguration === undefined) {
```

Comment:
This should probably have a tri-state with an `UNITIALIZED` symbol so we don't end up attempting to load it multiple times.

Reply:
Applies to the Webpack 5 version too.